api/src/main/java/org/openmrs/CohortMembership.java

```diff
@@ -99,6 +120,20 @@ public void setEndDate(Date endDate) {
 
 	@Override
 	public int compareTo(CohortMembership o) {
-		return this.getPatient().getPatientId() - o.getPatient().getPatientId();
+		int ret = -1;
+		if (Objects.equals(this.getPatientId(), o.getPatientId())
+				&& Objects.equals(this.getCohort().getCohortId(), o.getCohort().getCohortId())
+				&& this.getStartDate().equals(o.getStartDate())
+				&& OpenmrsUtil.compare(this.getStartDate(), o.getStartDate()) == 0
+				&& ((this.getEndDate() != null && o.getEndDate() != null
+				&& OpenmrsUtil.compare(this.getEndDate(), o.getEndDate()) == 0)
+				|| (this.getEndDate() == null && o.getEndDate() == null))) {
+			ret = 0;
+		} else if (this.isActive() && !o.isActive()) {
+			ret = -1;
+		} else if (!this.isActive() && o.isActive()) {
+			ret = 1;
+		}
+		return ret;
```

Comment:
This should not return an arbitrary value, you need to change it to:

`return OpenmrsUtil.compare(this.startDate(), o.startDate());`